### PR TITLE
Store characters in database

### PR DESCRIPTION
**Pull request by hernanhorquin96**
In this PR i add a button to refresh characters and store them in realm database. Also i add repository contracts and service contracts.
In next PR i will add the unit tests for all presenters.

### CODE REVIEW

**Review by BQuinteros — APPROVED** [review]

**Review by mauroLonghi — APPROVED** [review]

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/data/service/StoreCharactersServiceImpl.kt [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+import com.puzzlebench.clean_marvel_kotlin.domain.contracts.CharacterRepository
+import com.puzzlebench.clean_marvel_kotlin.domain.model.Character
+import io.realm.Realm
+
+class StoreCharactersServiceImpl(private val mapper: CharacterMapperRealm = CharacterMapperRealm()) : CharacterRepository {
+
+    override fun saveCharactersDB(characters: List<Character>) {
+        val realm = Realm.getDefaultInstance()
+        realm.executeTransaction {
+            realm.deleteAll()
```
Comment (SartiMau) [inline app/src/main/java/com/puzzlebench/clean_marvel_kotlin/data/service/StoreCharactersServiceImpl.kt]: it is necessary the deleteAll here? instead cannot be an update method?

Comment (SartiMau) [reply app/src/main/java/com/puzzlebench/clean_marvel_kotlin/data/service/StoreCharactersServiceImpl.kt]: you can use the realm method **insertOrUpdate**

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/domain/model/CharacterRealm.kt [hunk 2]
```diff
@@ -0,0 +1,16 @@
+package com.puzzlebench.clean_marvel_kotlin.domain.model
+
+import io.realm.RealmObject
+import io.realm.annotations.PrimaryKey
+
+open class CharacterRealm(
+    @PrimaryKey var id: Int = DEFAULT_ID,
+    var name: String? = DEFAULT_STRING,
```
Comment (SartiMau) [inline app/src/main/java/com/puzzlebench/clean_marvel_kotlin/domain/model/CharacterRealm.kt]: it is necessary that the variable be nullable? pls check this.

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/domain/model/ThumbnailRealm.kt [hunk 3]
```diff
@@ -0,0 +1,12 @@
+package com.puzzlebench.clean_marvel_kotlin.domain.model
+
+import io.realm.RealmObject
+
+open class ThumbnailRealm(
+        var path: String? = DEFAULT_STRING,
```
Comment (SartiMau) [inline app/src/main/java/com/puzzlebench/clean_marvel_kotlin/domain/model/ThumbnailRealm.kt]: same

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/domain/usecase/StoreCharactersUseCase.kt [hunk 4]
```diff
@@ -0,0 +1,9 @@
+package com.puzzlebench.clean_marvel_kotlin.domain.usecase
+
+import com.puzzlebench.clean_marvel_kotlin.domain.contracts.CharacterRepository
+import com.puzzlebench.clean_marvel_kotlin.domain.model.Character
+
+
+open class StoreCharactersUseCase(private val storeCharactersService: CharacterRepository) {
```
Comment (SartiMau) [inline app/src/main/java/com/puzzlebench/clean_marvel_kotlin/domain/usecase/StoreCharactersUseCase.kt]: Why this is an open class and method?

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/base/BaseRealm.kt [hunk 5]
```diff
@@ -0,0 +1,12 @@
+package com.puzzlebench.clean_marvel_kotlin.presentation.base
+
+import android.app.Application
+import io.realm.Realm
+
+class BaseRealm() : Application() {
```
Comment (SartiMau) [inline app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/base/BaseRealm.kt]: rename as BaseRealmApplication pls and put this in the package root, in presentation not inside the base package

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/model/CharactersModel.kt [hunk 6]
```diff
@@ -2,9 +2,14 @@ package com.puzzlebench.clean_marvel_kotlin.presentation.mvp.model
...
+import com.puzzlebench.clean_marvel_kotlin.domain.usecase.StoreCharactersUseCase
 import com.puzzlebench.clean_marvel_kotlin.presentation.mvp.contracts.CharactersContract
 import io.reactivex.Observable
 
-class CharactersModel(private val getCharacterServiceUseCase: GetCharacterServiceUseCase): CharactersContract.Model {
+class CharactersModel(private val getCharacterServiceUseCase: GetCharacterServiceUseCase,
+                      private val storeCharactersUseCase: StoreCharactersUseCase
+) : CharactersContract.Model {
     override fun getCharacterServiceUseCase(): Observable<List<Character>> = getCharacterServiceUseCase.invoke()
+    override fun storeCharactersUseCase(characters: List<Character>) = storeCharactersUseCase.invoke(characters)
```
Comment (SartiMau) [inline app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/model/CharactersModel.kt]: the .invoke is not necessary you can call **storeCharactersUseCase(characters)**

**Review by SartiMau — CHANGES_REQUESTED** [review]

**Review by lucianoeli — APPROVED** [review]

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/data/mapper/CharacterMapperRealm.kt [hunk 7]
```diff
@@ -0,0 +1,36 @@
+package com.puzzlebench.clean_marvel_kotlin.data.mapper
+
+import com.puzzlebench.clean_marvel_kotlin.domain.model.Character
+import com.puzzlebench.clean_marvel_kotlin.domain.model.CharacterRealm
+import com.puzzlebench.clean_marvel_kotlin.domain.model.Thumbnail
+import com.puzzlebench.clean_marvel_kotlin.domain.model.ThumbnailRealm
+
+open class CharacterMapperRealm : BaseMapperRepository<CharacterRealm, Character> {
+
+    override fun transform(characterRealm: CharacterRealm): Character = Character(
```
Comment (nicolbidone) [inline app/src/main/java/com/puzzlebench/clean_marvel_kotlin/data/mapper/CharacterMapperRealm.kt]: Can this have a more descriptive name?

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/domain/usecase/StoreCharactersUseCase.kt (hunk 4 above)
Comment (nicolbidone) [reply app/src/main/java/com/puzzlebench/clean_marvel_kotlin/domain/usecase/StoreCharactersUseCase.kt]: This happends in both useCases, pls check it

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/data/mapper/CharacterMapperRealm.kt (hunk 7 above)
Comment (hernanhorquin96) [reply app/src/main/java/com/puzzlebench/clean_marvel_kotlin/data/mapper/CharacterMapperRealm.kt]: Done

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/base/BaseRealm.kt (hunk 5 above)
Comment (hernanhorquin96) [reply app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/base/BaseRealm.kt]: Done

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/domain/model/CharacterRealm.kt (hunk 2 above)
Comment (hernanhorquin96) [reply app/src/main/java/com/puzzlebench/clean_marvel_kotlin/domain/model/CharacterRealm.kt]: No, its not necessary. I will change this in next commit!

File: app/src/main/java/com/puzzlebench/clean_marvel_kotlin/data/service/StoreCharactersServiceImpl.kt (hunk 1 above)
Comment (hernanhorquin96) [reply app/src/main/java/com/puzzlebench/clean_marvel_kotlin/data/service/StoreCharactersServiceImpl.kt]: Okey, thanks Mau!

**Review by mauroLonghi — APPROVED** [review]

**Review by SartiMau — APPROVED** [review]

**Review by nicolbidone — APPROVED** [review]

**Pull request merged**